File: Rust/ShuntingYard/parser.rs
use crate::lexer::{Association, Token};

pub fn to_rpn(tokens: &[Token]) -> Result<Vec<Token>, String> {
    let mut out: Vec<Token> = Vec::new();
    let mut ops: Vec<Token> = Vec::new();
    let mut arities: Vec<usize> = Vec::new();

    let mut i = 0;
    use Token::*;
    while i < tokens.len() {
        match &tokens[i] {
            Number(n) => out.push(Number(*n)),
            Ident(name) => {
                if matches!(tokens.get(i + 1), Some(Token::LParen)) {
                    ops.push(Token::Ident(name.clone()));
                    arities.push(0);
                } else {
                    return Err(format!(
                        "Unknown identifier '{}' (not a function call)",
                        name
                    ));
                }
            },
            Operator(op1) => {
                while let Some(top) = ops.last() {
                    match top {
                        Operator(op2) => {
                            let cond = match op1.association() {
                                Association::Left => op2.prec() >= op1.prec(),
                                Association::Right => op2.prec() > op1.prec(),
                            };
                            if cond {
                                out.push(ops.pop().unwrap());
                            } else {
                                break;
                            }
                        },
                        _ => break,
                    }
                }
                ops.push(Token::Operator(*op1));
            },
            LParen => {
                ops.push(Token::LParen);
            },
            Comma => {
                while let Some(top) = ops.last() {
                    if matches!(top, Token::LParen) {
                        break;
                    }
                    out.push(ops.pop().unwrap());
                }
                if let Some(a) = arities.last_mut() {
                    *a += 1;
                } else {
                    return Err("Comma not inside a function call".into());
                }
            },
            RParen => {
                while let Some(top) = ops.last() {
                    if matches!(top, Token::LParen) {
                        break;
                    }
                    out.push(ops.pop().unwrap());
                }
                if !matches!(ops.pop(), Some(Token::LParen)) {
                    return Err("Mismatched parentheses".into());
                }
                if let Some(Token::Ident(name)) = ops.last().cloned() {
                    ops.pop();
                    let argc = if let Some(a) = arities.pop() {
                        a + 1
                    } else {
                        0
                    };
                    out.push(Token::Call { name, argc });
                }
            },
            Call { .. } => unreachable!(),
        }
        if !arities.is_empty() {
            match tokens[i] {
                Number(_) | Ident(_) | RParen => {},
                _ => {},
            }
        }
        i += 1;
    }

    while let Some(t) = ops.pop() {
        match t {
            LParen | RParen => return Err("Mismatched parentheses".into()),
            Ident(n) => {
                out.push(Call { name: n, argc: 0 });
            },
            _ => out.push(t),
        }
    }
    Ok(out)
}

File: Fsharp/ShuntingYard/Parser.fs
module Parser

open System.Collections.Generic
open Lexer

let to_rpn (tokens: token[]) =
    let mutable out = Queue<token>()
    let mutable ops = Stack<token>()

    for token in tokens do
        match token with
        | Number n -> out.Enqueue(Number n)
        | Operator o1 ->
            let mutable continueLooping = true

            while continueLooping && ops.Count > 0 do
                match ops.Peek() with
                | Operator o2 when
                    precedence o2 > precedence o1
                    || precedence o2 = precedence o1 && is_left_associative o1
                    ->
                    out.Enqueue(ops.Pop())
                | _ -> continueLooping <- false

            ops.Push(Operator o1)
        | LParen -> ops.Push LParen
        | RParen ->
            while ops.Count > 0 && ops.Peek() <> LParen do
                out.Enqueue(ops.Pop())

            if ops.Count > 0 then
                ops.Pop() |> ignore

    while ops.Count > 0 do
        out.Enqueue(ops.Pop())

    out.ToArray()
